### PR TITLE
Workspace level doc handling

### PR DESCRIPTION
**Pull request by davidanthoff**
Goes together with JuliaEditorSupport/julia-vscode#116.

### CODE REVIEW

**Comment by ZacLN** [comment]
How far off are we with this? I'm using part of this to enable the first stage of the handling of `include` calls

**Comment by davidanthoff** [comment]
I think I need to find maybe an hour or two to finish this. Hopefully this week, a bit tricky right now, end of semester etc.

**Comment by ZacLN** [comment]
Yep yep fair enough!

**Comment by davidanthoff** [comment]
@ZacLN Ready for review!

File: src/provider_misc.jl [hunk 1]
```diff
@@ -12,6 +12,18 @@ const serverCapabilities = ServerCapabilities(
 
 function process(r::JSONRPC.Request{Val{Symbol("initialize")},Dict{String,Any}}, server)
     server.rootPath=haskey(r.params,"rootPath") ? r.params["rootPath"] : ""
+    if server.rootPath!=""
+        for (root, dirs, files) in walkdir(server.rootPath)
+            for file in files
+                if splitext(file)[2]==".jl"
+                    filepath = joinpath(root, file)
+                    uri = string("file:///", replace(replace(filepath, '\\', '/'), ":", "%3A"))
```
Comment (ZacLN) [inline src/provider_misc.jl]: linux (and MacOS?) already use forward slashes so we need something along the lines of: 
`uri = is_windows() ? string("file:///", replace(replace(filepath, '\\', '/'), ":", "%3A")) : string("file://", filepath)
            println(uri)`

File: src/provider_misc.jl [hunk 2]
```diff
@@ -21,7 +33,13 @@ function JSONRPC.parse_params(::Type{Val{Symbol("initialize")}}, params)
...
 function process(r::JSONRPC.Request{Val{Symbol("textDocument/didOpen")},DidOpenTextDocumentParams}, server)
-    server.documents[r.params.textDocument.uri] = Document(r.params.textDocument.text)
+    uri = r.params.textDocument.uri
+    if !haskey(server.documents, uri)
+        server.documents[uri] = Document(uri, r.params.textDocument.text, false)
+    end
+    doc = server.documents[uri]
+    set_open_in_editor(doc, true)
+
     parseblocks(r.params.textDocument.uri, server)
```
Comment (ZacLN) [inline src/provider_misc.jl]: use `uri`?

**Comment by davidanthoff** [comment]
@ZacLN Thanks for the review, I think I've addressed everything now.

File: src/provider_misc.jl [hunk 3]
```diff
@@ -16,8 +16,8 @@ function process(r::JSONRPC.Request{Val{Symbol("initialize")},Dict{String,Any}},
         for (root, dirs, files) in walkdir(server.rootPath)
             for file in files
                 if splitext(file)[2]==".jl"
-                    filepath = joinpath(root, file)
-                    uri = string("file:///", replace(replace(filepath, '\\', '/'), ":", "%3A"))
+                    filepath = joinpath(root, file)
+                    uri = string("file:///", is_windows() ? replace(replace(filepath, '\\', '/'), ":", "%3A") : filepath)
```
Comment (ZacLN) [inline src/provider_misc.jl]: still one too many slashes, on linux the filepaths lead with a forward slash so we end up with `file:////blah/blah`

Comment (davidanthoff) [inline src/provider_misc.jl]: Ah, bugger. Ok, I've just pushed a new version, hopefully that fixes this for good.

**Pull request merged**